File: Excel/600-699/689/689 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/689 Consecutive Numbers Marking.xlsx"
input = read_excel(path, range = "A1:B29")
test  = read_excel(path, range = "D1:D29")

result = input %>%
  mutate(rn = row_number()) %>%
  group_by(
    group_id = cumsum(ID != lag(ID, default = first(ID)) | 
                        (Number - rn) != lag(Number - rn, default = first(Number - rn)))
  ) %>%
  mutate(answer_expected = n()) %>%
  ungroup() %>%
  select(answer_expected)

all.equal(result$answer_expected, test$`Answer Expected`)
# TRUE

File: Excel/600-699/689/689 Challenge.py
import pandas as pd

path = "689 Consecutive Numbers Marking.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=29)
test = pd.read_excel(path, usecols="D", nrows=29)

input['rn'] = range(1, len(input) + 1)
input['group_id'] = (
    (input['ID'] != input['ID'].shift(1)) |
    ((input['Number'] - input['rn']) != (input['Number'] - input['rn']).shift(1))
).cumsum()
input['answer_expected'] = input.groupby('group_id')['group_id'].transform('size')
result = input[['answer_expected']]

print(result['answer_expected'].equals(test['Answer Expected']))
